Content.Client/GameObjects/EntitySystems/DragDropSystem.cs

```diff
@@ -0,0 +1,402 @@
+        private float _mouseDownTime;
+        // screen pos where the mouse down began
+        private Vector2 _mouseDownScreenPos;
+        // how much time since last recheck of all possible targets
+        private float _targetRecheckTime;
+        // reserved initial mousedown event so we can replay it if no drag ends up being performed
+        private PointerInputCmdHandler.PointerInputCmdArgs? _savedMouseDown;
+        // whether we are currently replaying the original mouse down, so we
+        // can ignore any events sent to this system
+        private bool _isReplaying;
+
+        private ShaderInstance _dropTargetInRangeShader;
+        private ShaderInstance _dropTargetOutOfRangeShader;
+        private SharedInteractionSystem _interactionSystem;
+        private InputSystem _inputSystem;
+
+        private List<SpriteComponent> highlightedSprites = new List<SpriteComponent>();
+
+        private enum DragState
+        {
+            NOT_DRAGGING,
+            // not dragging yet, waiting to see
+            // if they hold for long enough
+            MOUSEDOWN,
+            // currently dragging something
+            DRAGGING,
+        }
+
+
+        public override void Initialize()
+        {
+            base.Initialize();
+            IoCManager.InjectDependencies(this);
+            _state = DragState.NOT_DRAGGING;
+
+            _dropTargetInRangeShader = _prototypeManager.Index<ShaderPrototype>(ShaderDropTargetInRange).Instance();
+            _dropTargetOutOfRangeShader = _prototypeManager.Index<ShaderPrototype>(ShaderDropTargetOutOfRange).Instance();
+            _interactionSystem = EntitySystem.Get<SharedInteractionSystem>();
+            _inputSystem = EntitySystem.Get<InputSystem>();
+            // needs to fire on mouseup and mousedown so we can detect a drag / drop
+            CommandBinds.Builder
+                .Bind(EngineKeyFunctions.Use, new PointerInputCmdHandler(OnUse, false))
+                .Register<DragDropSystem>();
+
+        }
+
+        public override void Shutdown()
+        {
+            CancelDrag(false);
+            CommandBinds.Unregister<DragDropSystem>();
+            base.Shutdown();
+        }
+
+        private bool OnUse(in PointerInputCmdHandler.PointerInputCmdArgs args)
+        {
+            // not currently predicted
+            if (_inputSystem.Predicted) return false;
+
+            // currently replaying a saved click, don't handle this because
+            // we already decided this click doesn't represent an actual drag attempt
+            if (_isReplaying) return false;
+
+            if (args.State == BoundKeyState.Down)
+            {
+                return OnUseMouseDown(args);
+            }
+            else if (args.State == BoundKeyState.Up)
+            {
+                return OnUseMouseUp(args);
+            }
+
+            return false;
+        }
+
+        private bool OnUseMouseDown(PointerInputCmdHandler.PointerInputCmdArgs args)
```

Comment:
Tiny perf optimization here, pass this as `in` to reduce expensive stack copying (since that is a huge struct).

```suggestion
        private bool OnUseMouseDown(in PointerInputCmdHandler.PointerInputCmdArgs args)
```